src/main/java/sparksoniq/jsoniq/compiler/translator/expr/primary/FunctionCall.java

```diff
@@ -52,11 +62,64 @@ public String getFunctionName() {
 
     @Override
     public List<ExpressionOrClause> getDescendants(boolean depthSearch) {
-        List<ExpressionOrClause> result = new ArrayList<>();
-        result.addAll(this._arguments);
+        List<ExpressionOrClause> result = new ArrayList<>(this._arguments);
         return getDescendantsFromChildren(result, depthSearch);
     }
 
+    @Override
+    public final void initHighestExecutionMode() {
+        throw new OurBadException("Function call expressions do not use the highestExecutionMode initializer");
+    }
+
+    public void initFunctionCallHighestExecutionMode(boolean ignoreMissingFunctionError) {
+        FunctionIdentifier identifier = new FunctionIdentifier(this._functionName, this._arguments.size());
+        if (Functions.checkBuiltInFunctionExists(identifier)) {
```

Comment:
My intuition tells me that this code could be simplified by adding a return statement after each _highestExecutionMode assignments. Then all elses would go away and the structure would become flatter and easier to read.

Reply:
Offline discussion: function catalogue maps each builtin function to execution mode, and here we only look it up.

Reply:
I agree with the simplification using returns and I will implement this.

Regarding the second point: for the case of LocalRuntimeIterator, RDDRuntimeIterators or DataFrameRuntimeIterators where execution mode is always a fixed value, the catalogue approach would work. However in the case of HybridRuntimeIterators we need some custom logic in place that determines the execution mode based on the the child expression. But even this logic is not shared by all hybrid iterators due to peculiarities of individual iterators such as object 'keys', where we resort to local execution also when child is a DataFrame. 

Given this information that not all built-in functions have a static executionMode that can be accessed with a simple lookup, how would you like to proceed?

Reply:
The return oriented implementation is in place. I have also attempted to tidy up the code a bit to increase readability. 

I am looking forward to your feedback for further improvements or alternative solutions to the mentioned issue.

Reply:
Many thanks for the clean return-based implementation.

I am optimistic on the hybrid function iterators, especially looking at the cleaned code below, because there are not so many cases and we can come up with an larger list of function-specific execution modes (local, RDD, DF, but also same-as-first-parameter, etc). Then we just need to map this larger list (stored and looked up in the catalogue) to our "regular" execution mode list looking at the parameter iterators.

Could you please prepare a list of all functions that are based on a hybrid iterator? Then we'll take a closer look at them.

Thanks!

Reply:
These are the Hybrid function iterators:
- keys
- values
- distinct-values

- json-file
- parallelize
- text-file

Do you wish to perform this refactoring within this PR or shall we address introduction of function executionMode catalogue and functionIterator-only execution modes in a separate PR?

Reply:
Is it not just keys, values and distinct-values?

Among these: json-file, parallelize and text-file are anyway always RDDs (RDDRuntimeIterator), correct?

So if it is just keys, values and distinct-values that are "truly" hybrid and could be either local or RDD, then it should be quite easy as we only need one more case: when the highestExecutionMode is taken from the first parameter's highestExecutionMode. All three functions work that way.

This can be done with a separate, function-specific enum.

Reply:
Yes absolutely, that was a mistake. It's only the three that you have mentioned.

The cases are accurate. I will implement these next.